Plugins/Sycl/include/Acts/Plugins/Sycl/Seeding/Seedfinder.hpp

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright (C) 2020 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+// Acts include(s).
+#include "Acts/Plugins/Sycl/Seeding/DeviceExperimentCuts.hpp"
+#include "Acts/Plugins/Sycl/Seeding/detail/Types.h"
+#include "Acts/Seeding/InternalSeed.hpp"
+#include "Acts/Seeding/InternalSpacePoint.hpp"
+#include "Acts/Seeding/SeedFilter.hpp"
+#include "Acts/Seeding/SeedfinderConfig.hpp"
+
+/// Forward declaration of incomplete type cl::sycl::queue
+inline namespace cl {
+namespace sycl {
+class queue;
+}
+};  // namespace cl
+
+namespace Acts::Sycl {
+
+/// @brief Seedfinding algorithm implemented with SYCL to offload computations
+/// to GPUs.
+void createSeedsForGroupSycl(
+    const std::shared_ptr<cl::sycl::queue>& q,
+    const detail::deviceSeedfinderConfig& configData,
+    const DeviceExperimentCuts& deviceCuts,
+    const std::vector<detail::deviceSpacePoint>& bottomSPs,
+    const std::vector<detail::deviceSpacePoint>& middleSPs,
+    const std::vector<detail::deviceSpacePoint>& topSPs,
+    std::vector<std::vector<detail::SeedData>>& seeds);
+
+/// @brief This function creates the SYCL queue object.
+///
+/// SYCL implementation details are hidden from this class, that's why we're
+/// creating the queue in a standalone translation unit.
+///
+/// @return A pointer to the queue.
+std::shared_ptr<cl::sycl::queue> createQueue(const std::string&);
+
+template <typename external_spacepoint_t>
+class Seedfinder {
+ public:
+  Seedfinder(Acts::SeedfinderConfig<external_spacepoint_t> config,
+             const Acts::Sycl::DeviceExperimentCuts& cuts,
+             const std::string& device_name_substring = "");
+
+  ~Seedfinder() = default;
+  Seedfinder() = delete;
+  Seedfinder(const Seedfinder<external_spacepoint_t>&) = delete;
+  Seedfinder<external_spacepoint_t>& operator=(
+      const Seedfinder<external_spacepoint_t>&) = delete;
+
+  /// Create all seeds from the space points in the three iterators.
+  /// Can be used to parallelize the seed creation
+  /// @param bottom group of space points to be used as innermost SP in a seed.
+  /// @param middle group of space points to be used as middle SP in a seed.
+  /// @param top group of space points to be used as outermost SP in a seed.
+  /// Ranges must return pointers.
+  /// Ranges must be separate objects for each parallel call.
+  /// @return vector in which all found seeds for this group are stored.
+  template <typename sp_range_t>
+  std::vector<Seed<external_spacepoint_t>> createSeedsForGroup(
+      sp_range_t bottomSPs, sp_range_t middleSPs, sp_range_t topSPs) const;
+
+ private:
+  Acts::SeedfinderConfig<external_spacepoint_t> m_config;
+  Acts::Sycl::DeviceExperimentCuts m_deviceCuts;
+  std::shared_ptr<cl::sycl::queue> m_queue;
```

Comment:
:confused: I find it curious that [std::shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr) works correctly when a given compilation unit only sees a forward declaration of the type that it uses...

I think this is a pretty fragile/platform dependent setup. :frowning: As we discussed before, I think it would be much safer to handle the queue through a wrapper class like:

https://github.com/krasznaa/acts/blob/cuda-improvements-20200826/Plugins/Cuda/include/Acts/Plugins/Cuda/Utilities/StreamWrapper.hpp

This is not a huge priority to change, I just don't like shared pointers all that much to begin with. (Though I know that Acts uses them extensively...)

Reply:
I think the fact that `shared_ptr` works with forward declarations is due to careful implementation. Not arguing that `shared_ptr` should be used here (haven't looked at the code closely), but we generally try to be careful where we use them, and only do so if we feel like we get enough value out of them. Some code has `shared_ptr` for historical reasons, which is going to shrink with #412 for example.

Reply:
So I guess the question is, whether we will call the correct destructor to the queue object (which is an incomplete type to the Seedfinder class) when we destruct the Seedfinder object.

Some interesting resources about this:
https://howardhinnant.github.io/incomplete.html
https://stackoverflow.com/questions/2517245/why-really-deleting-an-incomplete-type-is-undefined-behaviour
https://stackoverflow.com/questions/25405135/shared-ptr-destructor-copy-and-incomplete-type

Let's look at an example here: https://wandbox.org/permlink/POjoXf3mt83bbW2k.
I guess I'm doing the same in the real code, but then it means it shouldn't work. (I'm confused, whether I should return the shared_ptr by reference.)

(I was thinking about providing some arguments about the topic, but I ended up so confused, I'd rather say nothing. I think we will call the correct destructor in case our object doesn't get destroyed prematurely.)

I guess we'll end up at rather not using shared_ptr, so I will change the implementation, and use a wrapper object instead (or call an external function from the destructor the deletes the pointer?).

Reply:
So I guess the question is, whether we will call the correct destructor to the queue object (which is an incomplete type to the Seedfinder class) when we destruct the Seedfinder object.

Some interesting resources about this:
https://howardhinnant.github.io/incomplete.html
https://stackoverflow.com/questions/2517245/why-really-deleting-an-incomplete-type-is-undefined-behaviour
https://stackoverflow.com/questions/25405135/shared-ptr-destructor-copy-and-incomplete-type

Let's look at an example here: https://wandbox.org/permlink/m0sjZv6jvXmeKtEd.

After all, I think we will call the correct destructor. The shared_ptr works with forward declarations, because all it stores is two pointers, and the "control block" has the destructor information (and other things).

Still, I'm not sure whether we should use this, I can just change the implementation.

(bonus: A [similar topic](https://feherenfekete.wordpress.com/2014/04/24/shared_ptr-polymorphic-magic-pitfall/) I heard a presentation about half a year ago, and I think I understood just now.)